common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/LastRequestTelemetry.java

```diff
@@ -76,10 +79,22 @@ void resetSilentSuccessCount() {
 
 
     void appendFailedRequest(final String apiId, final String correlationId, final String error) {
-        failedRequests.add(new FailedRequest(apiId, correlationId, error));
+        appendFailedRequest(new FailedRequest(apiId, correlationId, error));
     }
 
     void appendFailedRequest(final FailedRequest failedRequest) {
+        // this should usually not be greater than - at most should be equal to the cap
+        // because the only we to add to this list is via this append method.
+        // The only time this could be greater than the cap is for some existing devices that may
+        // have caught themselves in a bad state after accumulating too much telemetry in
+        // Shared Preferences. (of course prior to the cap being put in place).
+        // So will just take the last (most recent) 100 items here to get those out of the bad state,
+        // and also to avoid having too much telemetry in the cache going forward.
+        if (failedRequests.size() >= FAILED_REQUEST_CAP) {
+            final int beginIndex = failedRequests.size() - FAILED_REQUEST_CAP + 1;
+            final int endIndex = failedRequests.size();
+            failedRequests = failedRequests.subList(beginIndex, endIndex);
```

Comment:
I'm not sure at all that this is what we want to do.  subList returns a view on the original list.  So, as I read it, over time you'll end up with an ever expanding chain of references to new sublists (see SubList.subList).

https://cs.android.com/android/platform/superproject/+/master:libcore/ojluni/src/main/java/java/util/ArrayList.java;drc=6bd07db6e4f0806b658fc44bfee5dbef2c409540;l=1007

If we're only doing 100 of these, it might be better to just declare your own array and use it like a circular buffer, even though it means you'll have more complex code when somebody requests a view.

Reply:
Yeah, I would say let's keep the code simple for now. What do you say?

Reply:
Synced offline - @AdamBJohnsonx will write up a circular buffer and we can pick it up in the future.